cachi2/core/package_managers/gomod.py

```diff
@@ -878,27 +903,125 @@ def go_list_deps(pattern: Literal["./...", "all"]) -> Iterator[ParsedPackage]:
         cmd = [*go_list, "-deps", "-json=ImportPath,Module,Standard,Deps", pattern]
         return map(ParsedPackage.model_validate, load_json_stream(go(cmd, run_params)))
 
-    package_modules = (
+    package_modules = [
         module for pkg in go_list_deps("all") if (module := pkg.module) and not module.main
-    )
-
+    ]
+    package_modules.extend(workspace_modules)
     all_modules = _deduplicate_resolved_modules(package_modules, downloaded_modules)
 
     log.info("Retrieving the list of packages")
     all_packages = list(go_list_deps("./..."))
 
     _validate_local_replacements(all_modules, app_dir)
 
-    return ResolvedGoModule(main_module, all_modules, all_packages, modules_in_go_sum)
+    return ResolvedGoModule(main_module, all_modules, all_packages, modules_in_go_sum, go_work_root)
+
+
+ModuleDict = dict[str, Any]
+
+
+def _process_modules_json_stream(
+    app_dir: RootedPath, modules_json_stream: str
+) -> tuple[ModuleDict, list[ModuleDict]]:
```

Comment:
IMO, that will make the output a little less clear, since there's no way to tell that the first module is always the main module unless you look at this function's docstring or implementation. By returning a tuple, it is at least implicit that the first tuple element is unique somehow when compared to the elements in the list.

Reply:
That is IMO an inferior approach from design POV. If an element is special it needs to be reflected in the docs instead of trying to figure out a creative way how such a semantics needs to be represented in a data type.

Reply:
My reasoning is that we seldom read a function's docstring, and it is even less likely to be the first thing we do. By separating the results, we at least stablish that the objects returned are not the same from a business logic perspective, even if they belong from the same class.

But this is subjective, so unless you feel very strongly we should not use this design, I'd rather leave it as is. Btw, this is also how the newly introduced `_parse_local_modules` function is designed :angel:

Reply:
> Btw, this is also how the newly introduced _parse_local_modules function is designed 👼

The fact that both are introduced as part of this PR doesn't really make it any more correct :).
However, based on our recent discussion and more specifically based on this paragraph:

> The main problem that we reached is how to work around the need of having what Cachi2 defines as a "main module", since for each of the input packages fed to the CLI, a call to _resolve_gomod is made where the package is defined as the "main module" for purposes of the resolve algorithm. If the input package points to a folder that is not a valid go module, than we can't use it as the "main module".

I think we can keep it as is, since there's already the possibility that the main module concept may need to be revised in the future in which case many functions will have to be reworked, including this one. Despite still being in disagreement with the function signature, it should not be a show-stopper for this work.

Reply:
Mostly agree with Erik here, calling it a main module is false advertising. You're picking one arbitrary workspace module and calling it "main" (which probably has consequences for detecting the versions of the other workspace modules and other locally replaced modules)

But proper version detection would probably need bigger changes, this is a good enough start